### PR TITLE
[ComInterfaceGenerator] Warn if StringMarshalling doesn't match base and warn if base interface cannot be generated

### PR DESCRIPTION
**Pull request by jtschuster**
Warns if the base interface does not have the same StringMarshalling or if a base interface cannot be generated. Adds a LocationInfo to be able to produce diagnostics for interfaces without the ISymbol or anything that ruins incrementality.

This defines a few diagnostics / behaviors:
- If a base interface fails to be generated, then the inheriting interfaces will not be generated.
- If a base interface defines StringMarshalling, then all inheriting interfaces must match the base StringMarshalling.
- If a base interface does not define StringMarshalling (because it doesn't use strings or uses MarshaAs / MarshalUsing for each), an inheriting interface may define StringMarshalling with no error.

### CODE REVIEW

File: src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Unit.Tests/CompileFails.cs [hunk 1]
```diff
@@ -95,6 +95,218 @@ public static IEnumerable<object[]> InvalidUnmanagedToManagedCodeSnippetsToCompi
...
+            (StringMarshalling, Type?) customWithNoType = (StringMarshalling.Custom, null);
+            (StringMarshalling, Type?) utf8WithType = (StringMarshalling.Utf8, typeof(Utf16StringMarshaller));
+            DiagnosticResult[] emptyDiagnostics = new DiagnosticResult[] { };
+
+            // StringMarshalling.Custom / StringMarshallingCustomType invalid
+            yield return new object[]
+            {
+                ID(),
+                codeSnippets.DerivedWithStringMarshalling(customWithNoType),
+                new DiagnosticResult[] { new DiagnosticResult(GeneratorDiagnostics.InvalidStringMarshallingConfigurationOnInterface).WithLocation(0).WithArguments("Test.IStringMarshalling0", CustomStringMarshallingWithNoCustomTypeMessage) }
```
Comment (jkoritzinsky) [inline src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Unit.Tests/CompileFails.cs:117]:
```suggestion
                new[] {
                    VerifyComInterfaceGenerator.Diagnostic(GeneratorDiagnostics.InvalidStringMarshallingConfigurationOnInterface)
                        .WithLocation(0)
                        .WithArguments("Test.IStringMarshalling0", CustomStringMarshallingWithNoCustomTypeMessage) }
```

File: src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Unit.Tests/CompileFails.cs [hunk 2]
```diff
@@ -95,6 +95,218 @@ public static IEnumerable<object[]> InvalidUnmanagedToManagedCodeSnippetsToCompi
...
+            };
+            yield return new object[]
+            {
+                ID(),
+                codeSnippets.DerivedWithStringMarshalling(utf8WithType),
+                new DiagnosticResult[] { new DiagnosticResult(GeneratorDiagnostics.InvalidStringMarshallingConfigurationOnInterface).WithLocation(0).WithArguments("Test.IStringMarshalling0", CustomTypeSpecifiedWithNoStringMarshallingCustom) }
```
Comment (jkoritzinsky) [inline src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Unit.Tests/CompileFails.cs:123]:
```suggestion
                new[] {
                    VerifyComInterfaceGenerator.Diagnostic(GeneratorDiagnostics.InvalidStringMarshallingConfigurationOnInterface)
                        .WithLocation(0)
                        .WithArguments("Test.IStringMarshalling0", CustomTypeSpecifiedWithNoStringMarshallingCustom) }
```

File: src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Unit.Tests/CompileFails.cs [hunk 3]
```diff
@@ -95,6 +95,218 @@ public static IEnumerable<object[]> InvalidUnmanagedToManagedCodeSnippetsToCompi
...
+            //        [GeneratedComInterface(StringMarshalling = StringMarshalling.Utf16)]
+            //        [Guid("0E7204B5-4B61-5E06-B872-82BA652F2ECA")]
+            //        internal partial interface IStringMarshalling : INoStringMarshalling
+            //        {
+            //            public string GetString2();
+            //            public void SetString2(string value);
+            //        }
+            //    }
+            //    """;
+            //yield return new object[] { ID(), source, emptyDiagnostics };
```
Comment (jkoritzinsky) [inline src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Unit.Tests/CompileFails.cs:280]: What's the issue with this test? Can we either fix this or remove the commented-out code?

File: src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Unit.Tests/CompileFails.cs [hunk 4]
```diff
@@ -95,6 +95,218 @@ public static IEnumerable<object[]> InvalidUnmanagedToManagedCodeSnippetsToCompi
...
+
+            // Base many levels up fails, all inheriting fail
+            yield return new object[]
+            {
+                ID(),
+                codeSnippets.DerivedWithStringMarshalling(customWithNoType, customUtf16Marshalling, customUtf16Marshalling, customUtf16Marshalling, customUtf16Marshalling, customUtf16Marshalling),
+               new DiagnosticResult[] {new DiagnosticResult(GeneratorDiagnostics.InvalidStringMarshallingConfigurationOnInterface).WithLocation(0).WithArguments("Test.IStringMarshalling0", CustomStringMarshallingWithNoCustomTypeMessage) }.Concat(MismatchesWithLocations(1)).Concat(BaseCannotBeGeneratedWithLocations(2, 3, 4, 5)).ToArray()
```
Comment (jkoritzinsky) [inline src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Unit.Tests/CompileFails.cs:287]: Same transformation as above.

File: src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Unit.Tests/CompileFails.cs (hunk 3 above)
Comment (jtschuster) [reply src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Unit.Tests/CompileFails.cs:280]: Ah this was an issue with copying the parameter attributes as written in the source code to the unreachable exception stub. I'll push a fix to just avoid copying the attributes.

**Review by jkoritzinsky — APPROVED** [review]

**Pull request merged**